documents/model.go

```diff
@@ -754,3 +755,154 @@ func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []b
+	err = m.setCoreDocumentSalts()
+	if err != nil {
+		return errors.New("failed to generate CoreDocumentSalts")
+	}
+	return nil
+}
+
+// ATOwnerCanRead checks if the owner of the AT can read the document
+func (m *CoreDocumentModel) ATOwnerCanRead(tokenReq p2ppb.AccessTokenRequest , account identity.CentID) error {
+	// check if the access token is present in read rules
+	token, err := m.findAT(coredocumentpb.Action_ACTION_READ, tokenReq.AccessTokenId, func(role *coredocumentpb.Role, tokenID []byte) (*coredocumentpb.AccessToken, error){
+		return isATInRole(role, tokenReq.AccessTokenId)
+	})
+	if err != nil {
+		return err
+	}
+
+	granterID, err := identity.ToCentID(token.Granter)
+	if err != nil {
+		return err
+	}
+	// check that the token granter has the correct access type for the document
+	validGranter := m.AccountCanRead(granterID)
+	if !validGranter {
+		return errors.New("invalid access token granter")
+	}
+	// TODO:  check that the granter is the owner of the public key on the access token
+	
+	// check that the account requesting access is the grantee of the access token
+	if !bytes.Equal(account[:], token.Grantee) {
+		return errors.New("access token grantee is not the same as the account requesting access")
+	}
+	// validate the access token
+	err = validateAT(token.Key, token)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateAT validates that given access token against its signature
+// ethereum specific variation is currently false
+
+func validateAT (publicKey []byte, token *coredocumentpb.AccessToken) error {
+	tm := append(token.Identifier, token.Granter...)
+	tm = append(tm, token.Grantee...)
+	tm = append(tm, token.RoleIdentifier...)
+	tm = append(tm, token.DocumentIdentifier...)
+
+	validated := crypto.VerifyMessage(publicKey, tm, token.Signature, "CurveSecp256K1", true)
+	if !validated {
+		return errors.New("access token is invalid")
+	}
+	return nil
+}
+
+// isATInRole checks if the given access token is part of the core document role.
+func isATInRole(role *coredocumentpb.Role, tokenID []byte) (*coredocumentpb.AccessToken, error) {
+	for _, a := range role.AccessTokens{
+		if bytes.Equal(tokenID, a.Identifier) {
+			return a, nil
+		}
+	}
+	return nil, errors.New("access token not found")
+}
+
+// findAT calls OnRole for every role, returns the access token if it is found in the Roles of the document
+func (m *CoreDocumentModel) findAT (action coredocumentpb.Action, tokenID []byte, onRole func(role *coredocumentpb.Role, tokenID []byte) (*coredocumentpb.AccessToken, error)) (*coredocumentpb.AccessToken, error) {
+	cd := m.Document
+	var token coredocumentpb.AccessToken
+	for _, rule := range cd.ReadRules {
+		if rule.Action != action {
+			continue
+		}
+
+		for _, rk := range rule.Roles {
+			role, err := getRole(rk, cd.Roles)
+			if err != nil {
+				// seems like roles and rules are not in sync
+				// skip to next one
+				continue
+			}
+
+			at, err := onRole(role, tokenID)
```

Comment:
Instead of doing this call here, I think it's better to cache the roles that the tokenID is in before iterating over all the rules and then simply checking if it's in the list. 

```go
// Line 885
var tokenRoles := findRolesForAT(roles *coredocumentpb.Roles, tokenID []byte) 

// Line 899
for _, r := range(tokenRoles) {
    if r == role {
        // match
    }
}
```